docs/functionaldatatypes/zset.md

```diff
@@ -0,0 +1,276 @@
+---
+id: zset
+title: "ZSet"
+---
+
+`ZSet` is a version of a set that generalizes the measure of "how many times" each element appears in the set.
+
+Conceptually, we can think of a `ZSet[A, B]` as a `Map[A, B]`, where `A` represents the elements of the set and `B` represents some notion of how many times each element appears.
+
+We can then define a `Set` as a `ZSet` where the measure of how many times each element appears is a `Boolean`.
+
+```scala mdoc
+import zio.prelude.ZSet
+
+type Set[+A] = ZSet[A, Boolean]
+```
+
+That is, in a `Set` the measure of how many times an element appears is just yes or no. Either the element is in the set or it is not.
+
+However, we can also define many other variants of sets by choosing different types to measure how many times an element appears.
+
+For example, by using the `Natural` new type from ZIO Prelude, which describes integers greater than or equal to zero, we can define the type of a `MultiSet` or "bag" in some programming languages.
+
+```scala mdoc:reset
+import zio.prelude.ZSet
+import zio.prelude.newtypes.Natural
+
+type MultiSet[+A] = ZSet[A, Natural]
+```
+
+This ends up being a quite useful data type and can represent anything where the number of times an element appears matters but the order doesn't matter, such as a shopping cart.
+
+By using other measures of how many times an item appears we can describe other variants of sets such as `Int` for sets where items can appear a negative number of times or `Double` for sets where the number of times an item appears can be fractional.
+
+`ZSet` is focused on using the functional abstractions and new types in ZIO Prelude to describe operators on these generalizations of sets in as practical a way as possible.
+
+## Constructing ZSets
+
+The most basic way to construct a `ZSet` is from a `Map` where the keys in the map represent the elements in the `ZSet` and the values in the map represent the number of times each element appears.
+
+```scala mdoc
+val shoppingList: ZSet[String, Int] =
+  ZSet.fromMap(Map("apples" -> 2, "oranges" -> 3, "bananas" -> 5))
+```
+
+We can also construct a `ZSet` from existing collection types.
+
+If we have a normal `Set` from the Scala standard library, we can convert it to a `ZSet` using the `fromSet` operator.
+
+```scala mdoc
+val set: ZSet[Int, Boolean] =
+  ZSet.fromSet(Set(1, 2, 3))
+```
+
+If we have an `Iterable` we can convert it into a `ZSet` using the `fromIterable` constructor. The measure of how many times each item appears will just be how many times it appears in the `Iterable`.
+
+```scala mdoc
+val multiSet: MultiSet[Int] =
+ ZSet.fromIterable(List(1, 1, 2))
+```
+
+Finally, we can just construct an empty `ZSet` using the `empty` operator and add elements to it later.
```

Comment:
nitpick on "add elements to it later" as we're in a immutable data structure mindset, maybe some rewording could take place?
For example ... "and combine this empty `ZSet` with some non empty `ZSet` later"?

Reply:
```suggestion
Finally, we can just construct an empty `ZSet` using the `empty` operator and combine this empty ZSet with some non empty ZSet later.
```
(to avoid implying mutation)